### PR TITLE
Send SessionRecords to correct location

### PR DESCRIPTION
**Pull request by nisha-bhatia**
#15664

### CODE REVIEW

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 1]
```diff
@@ -78,7 +80,41 @@ private string GetSessionFilePath()
...
             string fileName = name + (IsAsync ? "Async" : string.Empty) + ".json";
-            return Path.Combine(TestContext.CurrentContext.TestDirectory,
+
+            var path = TestContext.CurrentContext.TestDirectory;
+            var newpath = path.Substring(0, path.IndexOf("artifacts")) + "sdk";
+
+            // check if Track 1 or Track 2
+            int firstIndex = path.IndexOf("bin\\") + "bin\\".Length;
+            int lastIndex = path.LastIndexOf(".Tests");
+            var testname = path.Substring(firstIndex, lastIndex - firstIndex);
```
Comment (bquantump) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:90]: Handle exceptions here if it substrings are not in the path. Just put the files in the old location if the path is not structured as expected, since the dll would be running in an unexpected location.

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 2]
```diff
@@ -78,7 +80,41 @@ private string GetSessionFilePath()
...
+            var rpname = "";
+            if (testname[0] == 'A') //Track 2
```
Comment (bquantump) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:92]: Confirm it is "Azure"

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 3]
```diff
@@ -78,7 +80,41 @@ private string GetSessionFilePath()
...
+            {
+                int rpnamefirstindex = testname.IndexOf("ResourceManager.") + "ResourceManager.".Length;
+                rpname = testname.Substring(rpnamefirstindex, testname.Length - rpnamefirstindex).ToLower();
+                rpname.Replace(".", "-");
+            }
+            else //Track 1
```
Comment (bquantump) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:98]: Confirm It is "Microsoft"

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 4]
```diff
@@ -78,7 +80,66 @@ private string GetSessionFilePath()
...
+            }
+
+            firstIndex += "bin\\".Length;
+            var testname = path.Substring(firstIndex, lastIndex - firstIndex);
+            var rpname = "";
+
+            if (testname.Substring(0,5) == "Azure") //Track 2
+            {
+                int rpnamefirstindex = testname.IndexOf("ResourceManager.") + "ResourceManager.".Length;
+                rpname = testname.Substring(rpnamefirstindex, testname.Length - rpnamefirstindex).ToLower();
```
Comment (bquantump) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:113]: Need to test if it management. Code should be if check if Azure.ResourceManager.

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 5]
```diff
@@ -78,7 +80,66 @@ private string GetSessionFilePath()
...
+                rpname.Replace(".", "-");
+            }
+            else if (testname.Substring(0,9) == "Management") //Track 1
+            {
+                int rpnamefirstindex = testname.IndexOf("Management.") + "Management.".Length;
```
Comment (bquantump) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:118]: Need to check if Microsoft.Managment.

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 6]
```diff
@@ -78,7 +80,66 @@ private string GetSessionFilePath()
             string className = testAdapter.ClassName.Substring(testAdapter.ClassName.LastIndexOf('.') + 1);
 
             string fileName = name + (IsAsync ? "Async" : string.Empty) + ".json";
-            return Path.Combine(TestContext.CurrentContext.TestDirectory,
+
+            var path = TestContext.CurrentContext.TestDirectory;
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:84]: This is an enhancement that we would want to enable for all Track 2 tests. It will be really helpful for users.

It might be better to use an attribute that we include in https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/Directory.Build.props.

```
  <ItemGroup>
    <AssemblyAttribute Include="System.Reflection.AssemblyMetadataAttribute">
    <_Parameter1>SourcePath</_Parameter1>
    <_Parameter2>$(MSBuildProjectDirectory)</_Parameter2>
    </AssemblyAttribute>
  </ItemGroup>
```

We can then retrieve the path like this from within RecordedTestBase:
```cs
string path = ((AssemblyMetadataAttribute) GetType().Assembly.GetCustomAttribute(typeof(AssemblyMetadataAttribute))).Value;
```

**Comment by JoshLove-msft** [comment]
We should remove the `UpdateSessionRecords` target and update all of our documentation that references it:
https://github.com/Azure/azure-sdk-for-net/search?q=UpdateSessionRecords

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs (hunk 6 above)
Comment (bquantump) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:84]: We were only going to do this for MGMT plane, are you suggesting we do this for data plane as well?

This suggestion seems simpler as well if we can get it from this property.

Comment (pakrym) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:84]: I think making this change both for track 1 and track 2 makes sense.

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 7]
```diff
@@ -78,7 +80,66 @@ private string GetSessionFilePath()
...
+            var testname = path.Substring(firstIndex, lastIndex - firstIndex);
+            var rpname = "";
+
+            if (testname.Substring(0,5) == "Azure") //Track 2
+            {
+                int rpnamefirstindex = testname.IndexOf("ResourceManager.") + "ResourceManager.".Length;
+                rpname = testname.Substring(rpnamefirstindex, testname.Length - rpnamefirstindex).ToLower();
+                rpname.Replace(".", "-");
+            }
+            else if (testname.Substring(0,9) == "Management") //Track 1
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:116]: This test framework is not used for track 1 projects.

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 8]
```diff
@@ -81,52 +81,42 @@ private string GetSessionFilePath()
...
+            var isMgmt = attributes["MSBuildProjectName"];
 
-            firstIndex += "bin\\".Length;
-            var testname = path.Substring(firstIndex, lastIndex - firstIndex);
+            string[] pathParts = path.Split("\\");
+            string testname = pathParts[pathParts.Length - 2];
             var rpname = "";
 
-            if (testname.Substring(0,5) == "Azure") //Track 2
+            if (testname.Substring(0,22) == "Azure.ResourceManager." && isMgmt == "true") //Track 2
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:98]: Why do we need to check if it is mgmt?

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 9]
```diff
@@ -81,52 +81,42 @@ private string GetSessionFilePath()
...
-            int lastIndex = path.LastIndexOf(".Tests");
-            if (firstIndex == -1 || lastIndex == -1)
+            Dictionary<string, string> attributes = new Dictionary<string, string>();
+            foreach (var attribute in customAttributes)
             {
-                return Path.Combine(TestContext.CurrentContext.TestDirectory,
-                "SessionRecords",
-                additionalParameterName == null ? className : $"{className}({additionalParameterName})",
-                fileName);
+                attributes.Add(((AssemblyMetadataAttribute) attribute).Key, ((AssemblyMetadataAttribute) attribute).Value);
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:89]: Why do we store all the key value pairs, rather than just grabbing the ones we need?

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 10]
```diff
@@ -6,7 +6,9 @@
 using System.Diagnostics;
 using System.IO;
 using System.Linq;
+using System.Reflection;
 using NUnit.Framework;
+using static System.Net.Mime.MediaTypeNames;
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:11]: Is this needed?

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 11]
```diff
@@ -78,7 +80,17 @@ private string GetSessionFilePath()
             string className = testAdapter.ClassName.Substring(testAdapter.ClassName.LastIndexOf('.') + 1);
 
             string fileName = name + (IsAsync ? "Async" : string.Empty) + ".json";
-            return Path.Combine(TestContext.CurrentContext.TestDirectory,
+
+            string path = ((AssemblyMetadataAttribute) GetType().Assembly.GetCustomAttribute(typeof(AssemblyMetadataAttribute))).Value;
+
+            if (path == null) // send to test directory
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:86]: Why do we need this fallback? The attribute will be in every project. If not populated, should we just throw?

**Comment by nisha-bhatia** [comment]
@JoshLove-msft @pakrym Microsoft.Azure.ServiceBus.UnitTests.API.ApiApprovals.ApproveAzureServiceBus test is failing because it is using the Assembly. Now the Assembly has an extra value of the SourcePath, and the Assembly cannot be changed. Do you guys know a workaround to this? Thanks.

Stack trace:
2020-10-20T19:47:16.0687670Z   X Microsoft.Azure.ServiceBus.UnitTests.API.ApiApprovals.ApproveAzureServiceBus [793ms]
2020-10-20T19:47:16.0689643Z   Error Message:
2020-10-20T19:47:16.0690522Z    Assert.Equal() Failure
2020-10-20T19:47:16.0691632Z                               â†“ (pos 20)
2020-10-20T19:47:16.0692954Z Expected: [assembly: System.Resources.NeutralResourcesLanguage("en-US")Â·Â·Â·
2020-10-20T19:47:16.0694237Z Actual:   [assembly: System.Reflection.AssemblyMetadata("SourcePath", "Â·Â·Â·
2020-10-20T19:47:16.0695358Z                               â†‘ (pos 20)
2020-10-20T19:47:16.0696127Z   Stack Trace:

**Comment by JoshLove-msft** [comment]
> @JoshLove-msft @pakrym Microsoft.Azure.ServiceBus.UnitTests.API.ApiApprovals.ApproveAzureServiceBus test is failing because it is using the Assembly. Now the Assembly has an extra value of the SourcePath, and the Assembly cannot be changed. Do you guys know a workaround to this? Thanks.
> 
> Stack trace:
> 2020-10-20T19:47:16.0687670Z X Microsoft.Azure.ServiceBus.UnitTests.API.ApiApprovals.ApproveAzureServiceBus [793ms]
> 2020-10-20T19:47:16.0689643Z Error Message:
> 2020-10-20T19:47:16.0690522Z Assert.Equal() Failure
> 2020-10-20T19:47:16.0691632Z â†“ (pos 20)
> 2020-10-20T19:47:16.0692954Z Expected: [assembly: System.Resources.NeutralResourcesLanguage("en-US")Â·Â·Â·
> 2020-10-20T19:47:16.0694237Z Actual: [assembly: System.Reflection.AssemblyMetadata("SourcePath", "Â·Â·Â·
> 2020-10-20T19:47:16.0695358Z â†‘ (pos 20)
> 2020-10-20T19:47:16.0696127Z Stack Trace:

Can we try to only add the attribute for test projects. I think there is a $(IsTestProject) property that we can use.

File: sdk/eventhub/Microsoft.Azure.EventHubs/Directory.Build.props [hunk 12]
```diff
@@ -8,6 +8,7 @@
   <PropertyGroup>
     <!-- Do not inherit implicit dependencies from the engineering system during build or packaging -->
     <ImportDefaultReferences>false</ImportDefaultReferences>
+    <IsTestProject>false</IsTestProject>
```
Comment (jsquire) [inline sdk/eventhub/Microsoft.Azure.EventHubs/Directory.Build.props:11]: Can you please help me understand why this change is being made?

**Review by jsquire — DISMISSED** [review]

**Comment by jsquire** [comment]
Please do not remove code owners from the review when their areas are being changed.

**Comment by nisha-bhatia** [comment]
I just ran the pipeline on master: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=588502&view=results. It's broken, could you guys please look into this? Thank you.

**Comment by JoshLove-msft** [comment]
> @pakrym @jsquire @JoshLove-msft I just ran the pipeline on master: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=588502&view=results. It's broken, could you guys please look into this? Thank you.

Most likely this is a merge issue. Can you try rebasing your changes on top of master?

**Comment by jsquire** [comment]
Looks like the `Azure.Core` solutions reference the old path for the Telemetry Exporter, before it was recently renamed. 

```text
##[error]Missing project. Solution references a project which does not exist: /home/vsts/work/1/s/sdk/core/Azure.Core/../../monitor/OpenTelemetry.Exporter.AzureMonitor/src/OpenTelemetry.Exporter.AzureMonitor.csproj. 
```

@pakrym - can you please advise how you think it should be addressed?

**Comment by nisha-bhatia** [comment]
@pakrym  I rebased to master, but the errors are still there. Thank you.

**Comment by pakrym** [comment]
There is a PR with a fix https://github.com/Azure/azure-sdk-for-net/pull/16224

**Comment by JoshLove-msft** [comment]
Looks good! Once we remove UpdateSessionRecords and update documentation/scripts, we should be good to go. Thanks @nisha-bhatia!

File: sdk/core/Azure.Core.TestFramework/README.md [hunk 13]
```diff
@@ -210,7 +210,7 @@ public class ConfigurationLiveTests: RecordedTestBase<AppConfigurationTestEnviro
 
 ## Recording
 
-When tests are run in recording mode, session records are saved to `artifacts/bin/<ProjectName>/<TargetFramework>/SessionRecords` directory. You can copy recordings to the project directory manually or by executing `dotnet msbuild /t:UpdateSessionRecords` in the test project directory.
+When tests are run in recording mode, session records are saved to the project directory manually in a folder named 'Session Records'.
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/README.md:213]:
```suggestion
When tests are run in recording mode, session records are saved to the project directory manually in a folder named 'SessionRecords'.
```

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/Record-Tests.ps1 [hunk 14]
```diff
@@ -27,7 +27,7 @@ else
     dotnet test --filter FullyQualifiedName~$TestName
 }
 
-dotnet msbuild /t:UpdateSessionRecords
+dotnet msbuild
```
Comment (pakrym) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/Record-Tests.ps1:30]: This is not required

**Review by pakrym — APPROVED** [review]

**Comment by pakrym** [comment]
https://github.com/Azure/azure-sdk-for-net/blob/master/eng/pipelines/templates/jobs/archetype-sdk-tests.yml#L145 needs updating as well.

File: eng/TestFramework.targets [hunk 15]
```diff
@@ -5,11 +5,6 @@
 
   <Target Name="AutoUpdateSessionRecords" Condition="'$(AutoUpdateSessionRecords)' == 'true'" DependsOnTargets="UpdateSessionRecordsInner" AfterTargets="Test;VSTest"/>
```
Comment (pakrym) [inline eng/TestFramework.targets:6]: This can be removed as well as `UpdateSessionRecordsInner` target.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/Record-Tests.ps1 (hunk 14 above)
Comment (drwill-ms) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/Record-Tests.ps1:30]: @vinagesh fyi

**Review by bquantump — APPROVED** [review]

File: sdk/iot/Azure.Iot.Hub.Service/tests/Record-Tests.ps1 [hunk 16]
```diff
@@ -27,7 +27,7 @@ else
     dotnet test --filter FullyQualifiedName~$TestName
 }
 
-dotnet msbuild /t:UpdateSessionRecords
+dotnet msbuild
```
Comment (pakrym) [inline sdk/iot/Azure.Iot.Hub.Service/tests/Record-Tests.ps1:30]: All of these dotnet msbuild invocations are not required

**Review by pakrym — APPROVED** [review]

**Review by JoshLove-msft — APPROVED** [review]

File: sdk/iot/Azure.Iot.Hub.Service/tests/UpdateSessionRecords.ps1 [hunk 17]
```diff
@@ -1 +1 @@
-﻿dotnet msbuild /t:UpdateSessionRecords
\ No newline at end of file
+﻿dotnet msbuild
```
Comment (pakrym) [inline sdk/iot/Azure.Iot.Hub.Service/tests/UpdateSessionRecords.ps1:1]: This file is not required as well.

**Comment by bquantump** [comment]
This appears to be a flakey test as rerunning the failed test allowed it to pass @pakrym 
Rerun: https://github.com/Azure/azure-sdk-for-net/runs/1318105030

2020-10-27T23:02:37.0825692Z   X AddsDelay [129ms]
2020-10-27T23:02:37.0825955Z   Error Message:
2020-10-27T23:02:37.0826248Z      Expected: less than or equal to 78
2020-10-27T23:02:37.0826554Z   But was:  80
2020-10-27T23:02:37.0826690Z 
2020-10-27T23:02:37.0827022Z   Stack Trace:
2020-10-27T23:02:37.0827411Z      at Azure.Search.Documents.Tests.ManualRetryDelayTests.AddsDelay() in /home/vsts/work/1/s/sdk/search/Azure.Search.Documents/tests/Batching/ManualRetryDelayTests.cs:line 29
2020-10-27T23:02:37.0828094Z 
2020-10-27T23:02:37.0828461Z   âˆš AddsMoreDelay [312ms]

**Comment by pakrym** [comment]
Yes, feel free to file failing tests. This one is already filed at https://github.com/Azure/azure-sdk-for-net/issues/16291
